CoreScriptsRoot/Modules/VR/Panel3D.lua

```diff
@@ -0,0 +1,760 @@
+	if self.isLookedAt then
+		self.transparency = 0
+		return
+	end
+	--Similarly, exit if we can't possibly see the panel.
+	if self.isOffscreen then
+		self.transparency = 1
+		return
+	end
+	--Otherwise, we'll want to calculate the transparency.
+	self.transparency = self:CalculateTransparency()
+end
+
+function Panel:Update(cameraCF, cameraRenderCF, userHeadCF, lookRay)
+	if self.forceShowUntilLookedAt and not self.part then
+		self:GetPart()
+		self:GetGUI()
+	end
+	if not self.part then
+		return
+	end
+
+	local isModal = (currentModal == self)
+	if not isModal and self.linkedTo and self.linkedTo == currentModal then
+		isModal = true
+	end
+	if currentModal and not isModal then
+		self:SetEnabled(false)
+		return
+	end
+
+	self:PreUpdate(cameraCF, cameraRenderCF, userHeadCF, lookRay)
+	if self.isVisible then
+		self:EvaluatePositioning(cameraCF, cameraRenderCF, userHeadCF)
+		self:EvaluateGaze(cameraCF, cameraRenderCF, userHeadCF, lookRay)
+
+		self:EvaluateTransparency(cameraCF, cameraRenderCF)
+	end
+end
+--End of Panel update methods
+
+--Panel virtual methods
+function Panel:PreUpdate() --virtual: handle positioning here
+end
+
+function Panel:OnUpdate() --virtual: handle transparency here
+end
+
+function Panel:OnMouseEnter(x, y) --virtual
+end
+
+function Panel:OnMouseLeave(x, y) --virtual
+end
+
+function Panel:OnEnabled(enabled) --virtual
+end
+
+function Panel:OnModalChanged(isModal) --virtual
+end
+
+function Panel:OnVisibilityChanged(visible) --virtual
+end
+
+function Panel:CalculateTransparency() --virtual
+	local guiWidth, guiHeight = self.gui.AbsoluteSize.X, self.gui.AbsoluteSize.Y
+	local lookX, lookY = self.lookAtPixel.X, self.lookAtPixel.Y
+
+	--Determine the distance from the edge; 
+	--if x is negative it's on the left side, meaning the distance is just absolute value
+	--if x is positive it's on the right side, meaning the distance is x minus the width
+	local xEdgeDist = lookX < 0 and -lookX or (lookX - guiWidth)
+	local yEdgeDist = lookY < 0 and -lookY or (lookY - guiHeight)
+	if lookX > 0 and lookX < guiWidth then
+		xEdgeDist = 0
+	end
+	if lookY > 0 and lookY < guiHeight then
+		yEdgeDist = 0
+	end
+	local edgeDist = math.sqrt(xEdgeDist ^ 2 + yEdgeDist ^ 2)
+
+	--since transparency is 0-1, we know how many pixels will give us 0 and how many will give us 1.
+	local offset = fullyOpaqueAtPixelsFromEdge
+	local interval = fullyTransparentAtPixelsFromEdge
+	--then we just clamp between 0 and 1.
+	return math.max(0, math.min(1, (edgeDist - offset) / interval))
+end
+--End of Panel virtual methods
+
+
+--Panel configuration methods
+function Panel:ResizeStuds(width, height, pixelsPerStud)
+	pixelsPerStud = pixelsPerStud or defaultPixelsPerStud
+
+	self.width = width
+	self.height = height
+
+	self.pixelScale = pixelsPerStud / defaultPixelsPerStud
+
+	local part = self:GetPart()
+	part.Size = Vector3.new(self.width * currentHeadScale, self.height * currentHeadScale, 1)
```

Comment:
I was wondering if we necessarily need to make the parts 1 stud thick or if we can make them thinner. If we already factor the thickness into the depth-position part of the GUI then it probably doesn't matter.